plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/FaultTolerantJdbcClient.java

```diff
@@ -0,0 +1,161 @@
+package io.trino.plugin.jdbc;
+
+import io.trino.plugin.jdbc.mapping.IdentifierMapping;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.security.ConnectorIdentity;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static io.trino.plugin.jdbc.JdbcWriteSessionProperties.getWriteBatchSize;
+import static io.trino.plugin.jdbc.JdbcWriteSessionProperties.isNonTransactionalInsert;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+
+public abstract class FaultTolerantJdbcClient
+        extends BaseJdbcClient {
+
+    public FaultTolerantJdbcClient(
+            BaseJdbcConfig config,
+            String identifierQuote,
+            ConnectionFactory connectionFactory,
+            QueryBuilder queryBuilder,
+            IdentifierMapping identifierMapping)
+    {
+        super(config, identifierQuote, connectionFactory, queryBuilder, identifierMapping);
+    }
+
+    public FaultTolerantJdbcClient(
+            String identifierQuote,
+            ConnectionFactory connectionFactory,
+            QueryBuilder queryBuilder,
+            Set<String> jdbcTypesMappedToVarchar,
+            IdentifierMapping identifierMapping)
+    {
+        super(identifierQuote, connectionFactory, queryBuilder, jdbcTypesMappedToVarchar, identifierMapping);
+    }
+
+    // TODO: Should this just be `return true`?
```

Comment:
I think we should just disallow  `NON_TRANSACTIONAL_INSERT` with FTE (task based retries).
Then the decision if rows deduplication should be used on insert or not, will be made solely base on `RetryMode` passed to `beginInsert` or `beginCreateTable`

Reply:
can you confirm that's all set?